posydon/binary_evol/DT/step_detached.py

```diff
@@ -599,188 +621,91 @@ def get_mist0(self, star, htrack):
             else:
                 initials = sol.x
         elif matching_method == "minimize":
+
+            def posydon_attribute(list_for_matching, star):
+                list_of_attributes = []
+                for attr in list_for_matching:
+                    list_of_attributes.append(getattr(star, attr))
+                return list_of_attributes
+
             if star.state in LIST_ACCEPTABLE_STATES_FOR_HMS:
                 # Initialezed ZAMS systems have no information of radius
-                if (star.log_R is None) or np.isnan(star.log_R):
+                if (star.log_R is None) or np.isnan(star.log_R): #TODO: This is no a good check for ZAMS!!
                     initials = (star.mass, 0)
                 else:
-                    MESA_label = ["mass", "center_h1", "log_R", "he_core_mass"]
-                    posydon_attribute = [star.mass, star.center_h1,
-                                         star.log_R, star.he_core_mass]
-                    rs = [20.0, 1.0, 2.0, 10.0]
-                    if self.verbose:
-                        print("Matching attributes and their normalizations :",
-                              MESA_label, rs)
-                    for i in MESA_label:
-                        if i not in self.root_keys:
-                            raise Exception("Expected matching parameter not "
-                                            "added in the MIST model options.")
-                    x0 = get_root0(MESA_label, posydon_attribute,
+                    list_for_matching = self.list_for_matching_HMS
+            elif star.state in LIST_ACCEPTABLE_STATES_FOR_postMS:
+                list_for_matching = self.list_for_matching_postMS
+                
+            elif star.state in LIST_ACCEPTABLE_STATES_FOR_HeStar:
+                list_for_matching = self.list_for_matching_HeStar
```

Comment:
Because for each kind of state the alternative matching require different `list_for_matching` and `htrack`, we create a function with input variable 'list_for matching' and 'htrack' to return the final solution 'sol'. 
And we call it here for different star state. If  np.abs(sol.fun) > tolerance_matching_integration or not sol.success, we change the list_for_matching and htrack.

Reply:
@kasdaglie Maybe we can implement the above suggestion from Zepei for the alternative matchings